libfreefare/mifare_desfire_key.c

```diff
@@ -124,16 +127,21 @@ mifare_desfire_key_get_version(MifareDESFireKey key)
 void
 mifare_desfire_key_set_version(MifareDESFireKey key, uint8_t version)
 {
+    if (key->type == T_AES) {
+	key->aes_version = version;
+	return;
+    }
+
     for (int n = 0; n < 8; n++) {
 	uint8_t version_bit = ((version & (1 << (7 - n))) >> (7 - n));
 	key->data[n] &= 0xfe;
 	key->data[n] |= version_bit;
 	if (key->type == T_DES) {
 	    key->data[n + 8] = key->data[n];
-	} else {
+	} else if (key->type == T_3DES) {
```

Comment:
If we start to have code `if (key->type == A) { ... } else if (key->type == B) { ... } else ...`, can you please turn this into a switch case without default like e.g. https://github.com/nfc-tools/libfreefare/blob/69bf99b33daf350d23aee8abffa96b4ed96151d1/libfreefare/mifare_desfire_crypto.c#L186-L195

That way, we have to have an explicit condition for each possible value, and if some new key type would be introduced in the future, the compiler would complain because all possible values are not handled in the switch statement.

Reply:
No problem, but how did you quote that other file inline in your comment like that?

Reply:
In this case, we really only care about DES keys, and hopefully there are no fancy new DES encryption types that will be introduced. I think that is a safe assumption. Forcing a warning for every newly added encryption type doesn't seem necessary in this block.

Reply:
I was wondering what was happening with 3K3DES… Correct me if I am wrong, but these manipulation only have the role of making a DES key stay a DES key and a 3DES key stay a 3DES key when the key version is changed; a problem the 3K3DES keys can not be affected by.

If it's the reason 3K3DES are not handled here, maybe adding a comment here may be enough: AES keys are handled at the top of the function and having to list them again seems tedious, I quite agree; and as you say, support for more types of keys seems improbable.

[oh, and to include a snippet in a comment, I just write the URL of a file with the relevant lines highlighted (the snippets title links to the URL I typed :stuck_out_tongue_winking_eye: )]